adapters/multiline/multiline.go

```diff
@@ -0,0 +1,245 @@
+	if err != nil {
+		return nil, errors.New("multiline: invalid value for MULTILINE_PATTERN (must be regexp): " + pattern)
+	}
+
+	matchType := os.Getenv("MULTILINE_MATCH")
+	if matchType == "" {
+		matchType = matchNonFirst
+	}
+	matchType = strings.ToLower(matchType)
+	matchFirstLine := false
+	negateMatch := false
+	switch matchType {
+	case matchFirst:
+		matchFirstLine = true
+		negateMatch = false
+	case matchLast:
+		matchFirstLine = false
+		negateMatch = false
+	case matchNonFirst:
+		matchFirstLine = true
+		negateMatch = true
+	case matchNonLast:
+		matchFirstLine = false
+		negateMatch = true
+	default:
+		return nil, errors.New("multiline: invalid value for MULTILINE_MATCH (must be one of first|last|nonfirst|nonlast): " + matchType)
+	}
+
+	flushAfter := 500 * time.Millisecond
+	flushAfterStr := os.Getenv("MULTILINE_FLUSH_AFTER")
+	if flushAfterStr != "" {
+		timeoutMS, err := strconv.Atoi(flushAfterStr)
+		if err != nil {
+			return nil, errors.New("multiline: invalid value for multiline_timeout (must be number): " + flushAfterStr)
+		}
+		flushAfter = time.Duration(timeoutMS) * time.Millisecond
+	}
+
+	parts := strings.SplitN(route.Adapter, "+", 2)
+	if len(parts) != 2 {
+		return nil, errors.New("multiline: adapter must have a sub-adapter, eg: multiline+raw+tcp")
+	}
+
+	originalAdapter := route.Adapter
+	route.Adapter = parts[1]
+	factory, found := router.AdapterFactories.Lookup(route.AdapterType())
+	if !found {
+		return nil, errors.New("bad adapter: " + originalAdapter)
+	}
+	subAdapter, err := factory(route)
+	if err != nil {
+		return nil, err
+	}
+	route.Adapter = originalAdapter
+
+	out := make(chan *router.Message)
+	checkInterval := flushAfter / 2
+
+	return &Adapter{
+		out:             out,
+		subAdapter:      subAdapter,
+		enableByDefault: enableByDefault,
+		pattern:         patternRegexp,
+		separator:       separator,
+		matchFirstLine:  matchFirstLine,
+		negateMatch:     negateMatch,
+		flushAfter:      flushAfter,
+		checkInterval:   checkInterval,
+		buffers:         make(map[string]*router.Message),
+		nextCheck:       time.After(checkInterval),
+	}, nil
+}
+
+// Stream sends log data to the next adapter
+func (a *Adapter) Stream(logstream chan *router.Message) {
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		a.subAdapter.Stream(a.out)
+		wg.Done()
+	}()
+	defer func() {
+		for _, message := range a.buffers {
+			a.out <- message
+		}
+
+		close(a.out)
+		wg.Wait()
+	}()
+
+	for {
+		select {
+		case message, ok := <-logstream:
+			if !ok {
+				return
+			}
+
+			if !multilineContainer(message.Container, a.enableByDefault) {
+				a.out <- message
+				continue
+			}
+
+			cID := message.Container.ID
+			old, oldExists := a.buffers[cID]
+			if a.isFirstLine(message) {
+				if oldExists {
+					a.out <- old
+				}
+
+				a.buffers[cID] = message
+			} else {
+				if oldExists {
+					old.Data += a.separator + message.Data
+					message = old
+				}
+
+				if a.isLastLine(message) {
```

Comment:
Here, you pass the whole buffered message to the `isLastLine()` function instead of only the current line, which makes the regex-matching faulty.
This caused my tests to fail using the nonlast-option. I changed it to the following:
```
is_last_line = a.isLastLine(message)

if oldExists {
   old.Data += a.separator + message.Data
   message = old
}

if is_last_line {
   a.out <- message
   if oldExists {
       delete(a.buffers, cID)
   }
```
That way, you always use only the current message for pattern matching but can still manipulate it in the first if-clause.
This works with all your tests and also with my test.

Reply:
Test-Case to reproduce my problem: 
```
{
			input: []string{
				"not yet",
				"Traceback",
				" tb1",
				" tb2!",
				"Error123",
				"no more traceback",
				"STATEMENT:",
				" still statement",
				"end of statement",
				"no more statement",
			},
			expected: []string{
				"not yet",
				"Traceback\n tb1\n tb2!\nError123",
				"no more traceback",
				"STATEMENT:\n still statement\nend of statement",
				"no more statement",
			},
			pattern:        regexp.MustCompile(`^(DETAIL:|STATEMENT:|Traceback|\s)`),
			matchFirstLine: false,
			negateMatch:    true,
		}
```

Reply:
shouldn't 
`isLastLine = a.isLastLine(message)` 
be 
`isLastLine := a.isLastLine(message)` 
?

Reply:
Sure thing, there's a typo which I don't have in the code repo

Reply:
opened a PR to masterada/logspout for that